packages/utils/crates/utils-types/src/sma.rs

```diff
@@ -61,35 +72,42 @@ where
     fn len(&self) -> usize {
         self.window.len()
     }
+
+    fn is_empty(&self) -> bool {
+        self.window.is_empty()
+    }
 }
 
-impl<T> NoSumSMA<T>
+impl<T, D> NoSumSMA<T, D>
 where
-    T: for<'a> Add<&'a T, Output = T> + Div<T, Output = T> + Clone + From<u32> + Display + Default,
+    T: Add<T, Output = T> + Div<D, Output = T> + Clone,
+    D: From<u32>,
 {
     /// Creates an empty SMA instance with the given window size.
     /// The maximum window size is u32::MAX and must be greater than 1.
     pub fn new(window_size: u32) -> Self {
         assert!(window_size > 1, "window size must be greater than 1");
         Self {
             window: AllocRingBuffer::new(window_size as usize),
+            _div: PhantomData,
         }
     }
 
     /// Creates SMA instance given window size and some initial samples.
     pub fn new_with_samples(window_size: u32, initial_samples: &[T]) -> Self {
         let mut ret = Self::new(window_size);
-        initial_samples.iter().for_each(|s| ret.add_sample(s.clone()));
+        initial_samples.iter().cloned().for_each(|s| ret.add_sample(s));
         ret
     }
 }
 
-impl<T> Display for NoSumSMA<T>
+impl<T, D> Display for NoSumSMA<T, D>
 where
-    T: for<'a> Add<&'a T, Output = T> + Div<T, Output = T> + Clone + From<u32> + Display + Default,
+    T: Add<T, Output = T> + Div<D, Output = T> + Clone + Default + Display,
+    D: From<u32>,
 {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        write!(f, "{}", self.get_average())
+        write!(f, "{}", self.get_average().unwrap_or_default())
```

Comment:
This would be more correct:
```suggestion
        write!(f, "{:?}", self.get_average())
```

Reply:
unwrap_or_default is more explicit IMHO

Reply:
For the log you would benefit from knowing whether the real value is 0.0 or it's just an alaias for None.